[#8241][feat] Support model_kwargs for pytorch backend

taylor-yb-lee opened the pull request:
<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description
Currently, pytorch backend does not support model_kwargs, so this PR adds model_kwargs in BaseLlmArgs and updates the new key values in the model_kwargs onto the LlmArgs recursively. 

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

taylor-yb-lee committed:
Support model_kwargs for torch backend

Signed-off-by: Taylor Yeonbok Lee <249374542+taylor-yb-lee@users.noreply.github.com>

taylor-yb-lee committed:
Applied review comment

Signed-off-by: Taylor Yeonbok Lee <249374542+taylor-yb-lee@users.noreply.github.com>

taylor-yb-lee committed:
Added unittest

Signed-off-by: Taylor Yeonbok Lee <249374542+taylor-yb-lee@users.noreply.github.com>

taylor-yb-lee committed:
Fix CI failure due to unregistered reference parameter for LLM API

Signed-off-by: Taylor Yeonbok Lee <249374542+taylor-yb-lee@users.noreply.github.com>

taylor-yb-lee commented:
/run bot

tensorrt-cicd commented:
[PR_Github #30478](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/30478/) [ run ] completed with state `SUCCESS`. Commit: `61dfccf` 
[/LLM/main/L0_MergeRequest_PR pipeline #23500](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/23500/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

taylor-yb-lee commented:
/bot --help

taylor-yb-lee commented:
/bot run

2ez4bz commented on tests/unittest/llmapi/test_llm_args.py:
```diff
@@ -138,6 +138,30 @@ def test_llm_args_with_pydantic_options(self):
+    num_hidden_layers: 2
+    """
+        dict_content = self._yaml_to_dict(yaml_content)
+        llm_args = TorchLlmArgs(model=llama_model_path)
```
Nit: the only difference between this test and the one above is the class definition being tested. You could consider parameterizing as below

```python
@pytest.mark.parametrize("llm_args_cls", [TrtLlmArgs, TorchLlmArgs])
def test_llm_args_with_model_kwargs(self, llm_args_cls):
    dict_content = self._yaml_to_dict(yaml_content)
    llm_args = llm_args_cls(model=llama_model_path)
    # etc.
```

taylor-yb-lee replied:
Thanks for the good tip!

taylor-yb-lee committed:
Applied review comment

Signed-off-by: Taylor Yeonbok Lee <249374542+taylor-yb-lee@users.noreply.github.com>

taylor-yb-lee commented on tensorrt_llm/_torch/model_config.py:
```diff
@@ -452,6 +452,57 @@ def cached_file(path_or_repo_id, file_name):
+        # Apply model_kwargs to override config parameters if provided
+        model_kwargs = kwargs.pop('model_kwargs', None)
+        if model_kwargs:
+            from tensorrt_llm.logger import logger
```
Oh this was redundant, removed. Thanks.

Superjomn reviewed (approved):
LGTM on the llmapi changes.

taylor-yb-lee commented:
/bot run

tensorrt-cicd commented:
[PR_Github #30769](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/30769/) [ run ] triggered by Bot. Commit: `3d589a1`

taylor-yb-lee committed:
Fix unittest failure

Signed-off-by: Taylor Yeonbok Lee <249374542+taylor-yb-lee@users.noreply.github.com>

taylor-yb-lee commented:
/bon run

tensorrt-cicd commented:
[PR_Github #31015](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31015/) [ run ] completed with state `SUCCESS`. Commit: `ebc00f1` 
[/LLM/main/L0_MergeRequest_PR pipeline #23966](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/23966/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

tensorrt-cicd commented:
[PR_Github #31107](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31107/) [ run ] triggered by Bot. Commit: `ebc00f1`

tensorrt-cicd commented:
[PR_Github #31107](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31107/) [ run ] completed with state `SUCCESS`. Commit: `ebc00f1` 
[/LLM/main/L0_MergeRequest_PR pipeline #24024](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/24024/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

tensorrt-cicd commented:
[PR_Github #31125](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31125/) [ run ] triggered by Bot. Commit: `ebc00f1`

taylor-yb-lee commented:
/bot run

tensorrt-cicd commented:
[PR_Github #31202](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31202/) [ run ] completed with state `SUCCESS`. Commit: `a93455c` 
[/LLM/main/L0_MergeRequest_PR pipeline #24112](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/24112/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

taylor-yb-lee commented:
/bot run --disable-fail-fast

taylor-yb-lee committed:
Fix unittest failure
- model_kwargs is an optional parameter, and it is not always set.
- This is handled in the standard ConfigLoader, but not in custom ConfigLoader such as MistralConfigLoader
- In current update, let model_kwargs be supported only by ConfigLoader and added warning in the custom loader

Signed-off-by: Taylor Yeonbok Lee <249374542+taylor-yb-lee@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #31343](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31343/) [ run ] completed with state `SUCCESS`. Commit: `72234a1` 
[/LLM/main/L0_MergeRequest_PR pipeline #24235](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/24235/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

tensorrt-cicd commented:
[PR_Github #31428](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31428/) [ run ] triggered by Bot. Commit: `72234a1`

tensorrt-cicd commented:
[PR_Github #31635](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31635/) [ run ] triggered by Bot. Commit: `be0949b`

tensorrt-cicd commented:
[PR_Github #31635](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31635/) [ run ] completed with state `FAILURE`. Commit: `be0949b` 
[/LLM/main/L0_MergeRequest_PR pipeline #24467 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/24467/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

tensorrt-cicd commented:
[PR_Github #31704](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31704/) [ run ] triggered by Bot. Commit: `be0949b`

taylor-yb-lee commented:
/bot run --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #31813](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31813/) [ run ] triggered by Bot. Commit: `be0949b`

tensorrt-cicd commented:
[PR_Github #31813](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31813/) [ run ] completed with state `SUCCESS`. Commit: `be0949b` 
[/LLM/main/L0_MergeRequest_PR pipeline #24626 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/24626/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

tensorrt-cicd commented:
[PR_Github #31838](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31838/) [ run ] triggered by Bot. Commit: `be0949b`

tensorrt-cicd commented:
[PR_Github #31914](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/31914/) [ run ] triggered by Bot. Commit: `278123f`

tensorrt-cicd commented:
[PR_Github #32022](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32022/) [ run ] triggered by Bot. Commit: `542acf8`

taylor-yb-lee commented:
/bot reuse-pipeline

taylor-yb-lee commented:
/bot run

tensorrt-cicd commented:
[PR_Github #32170](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32170/) [ run ] completed with state `SUCCESS`. Commit: `52cf348` 
[/LLM/main/L0_MergeRequest_PR pipeline #24943](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/24943/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

tensorrt-cicd commented:
[PR_Github #32182](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32182/) [ reuse-pipeline ] triggered by Bot. Commit: `52cf348`

taylor-yb-lee commented:
/bot run --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #32209](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32209/) [ run ] triggered by Bot. Commit: `52cf348`

taylor-yb-lee commented:
/bot reuse-pipeline

tensorrt-cicd commented:
[PR_Github #32247](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32247/) [ reuse-pipeline ] triggered by Bot. Commit: `52cf348`

tensorrt-cicd commented:
[PR_Github #32337](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32337/) [ run ] triggered by Bot. Commit: `5bd0079`

tensorrt-cicd commented:
[PR_Github #32337](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32337/) [ run ] completed with state `SUCCESS`. Commit: `5bd0079` 
[/LLM/main/L0_MergeRequest_PR pipeline #25057](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/25057/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

taylor-yb-lee commented:
/bot run

tensorrt-cicd commented:
[PR_Github #32362](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32362/) [ run ] completed with state `SUCCESS`. Commit: `5bd0079` 
[/LLM/main/L0_MergeRequest_PR pipeline #25078](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/25078/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

taylor-yb-lee commented:
/bot run --only-multi-gpu-test

tensorrt-cicd commented:
[PR_Github #32815](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32815/) [ run ] completed with state `SUCCESS`. Commit: `22f3e0a` 
[/LLM/main/L0_MergeRequest_PR pipeline #25390 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/25390/) completed with status: 'SUCCESS'

tensorrt-cicd commented:
[PR_Github #32882](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/32882/) [ run ] triggered by Bot. Commit: `22f3e0a`

tensorrt-cicd commented:
[PR_Github #33041](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/33041/) [ skip ] triggered by Bot. Commit: `22f3e0a`

tensorrt-cicd commented:
[PR_Github #33041](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/33041/) [ skip ] completed with state `SUCCESS`. Commit: `22f3e0a` 
Skipping testing for commit `22f3e0a`

2ez4bz commented:
![first try](https://media1.tenor.com/m/BQg5R2l0VMUAAAAC/lego-batman.gif)
